File: l4_inference/llm_service.py
# l4_inference/llm_service.py
import asyncio
import os
import json
import aiohttp

class LLMService:
    def __init__(self, provider="mock", api_key=None, base_url="https://api.siliconflow.cn/v1"):
        self.provider = provider
        self.api_key = api_key or os.getenv("SILICONFLOW_API_KEY")
        self.base_url = base_url

    async def generate(self, prompt, system_prompt="你是一个有用的助手"):
        """
        调用大模型生成文本
        """
        if self.provider.startswith("mock"):
            return await self._generate_mock()
        
        return await self._generate_api(prompt, system_prompt)

    async def _generate_mock(self):
        print(f"[L4] Generating response using {self.provider}...")
        await asyncio.sleep(1)
        return f"""
(Mock LLM Response from {self.provider})

**盈亏综述**
账户整体盈利，主要得益于重仓股表现良好。

**归因分析**
- 0700.HK: 腾讯近期发布财报超预期，加上游戏版号发放常态化，带动股价上涨。
- BABA: 阿里拆分消息落地，市场情绪回暖。

**风险提示**
当前持仓科技股占比较高，需注意板块回调风险。

**操作建议**
建议继续持有，可适当关注低估值的红利股以平衡风险。
"""

    async def _generate_api(self, prompt, system_prompt):
        print(f"[L4] Calling SiliconFlow API for model: {self.provider}...")
        
        if not self.api_key:
            return "[Error] API Key not found. Please set SILICONFLOW_API_KEY environment variable."

        url = f"{self.base_url}/chat/completions"
        headers = {
            "Content-Type": "application/json",
            "Authorization": f"Bearer {self.api_key}"
        }
        payload = {
            "model": self.provider,
            "messages": [
                {"role": "system", "content": system_prompt},
                {"role": "user", "content": prompt}
            ],
            "enable_thinking": False,
            "stream": True,
            "stream_options": {"include_usage": True} 
        }

        try:
            import time
            start_time = time.time()
            full_content = ""
            usage_info = None
            
            print(f"\n[L4 Stream Start - {self.provider}]\n")
            
            async with aiohttp.ClientSession() as session:
                async with session.post(url, headers=headers, json=payload) as response:
                    if response.status != 200:
                        error_text = await response.text()
                        return f"[Error] API request failed: {response.status} - {error_text}"
                    
                    async for line in response.content:
                        line = line.decode('utf-8').strip()
                        if line.startswith("data: ") and line != "data: [DONE]":
                            json_str = line[6:]  # Remove "data: "
                            try:
                                chunk = json.loads(json_str)
                                
                                # Handle usage info if present (often in the last chunk with empty choices)
                                if "usage" in chunk and chunk["usage"]:
                                    usage_info = chunk["usage"]
                                
                                # Handle content delta
                                if "choices" in chunk and chunk["choices"]:
                                    choice = chunk["choices"][0]
                                    delta = choice.get("delta", {})
                                    content_piece = delta.get("content", "")
                                    if content_piece:
                                        print(content_piece, end="", flush=True)
                                        full_content += content_piece
                            except json.JSONDecodeError:
                                pass
            
            end_time = time.time()
            duration = end_time - start_time
            print("\n") # Newline after stream
            
            # Print Stats
            if usage_info:
                total_tokens = usage_info.get("total_tokens", 0)
                completion_tokens = usage_info.get("completion_tokens", 0)
                # Note: SiliconFlow/OpenAI usage in stream might be just the delta or final total. 
                # Usually standard is final chunk has usage.
                
                # If we didn't get completion tokens from API, iterate to estimate or just show total
                tokens_generated = completion_tokens if completion_tokens > 0 else len(full_content) # Fallback to rough char count if 0
                
                speed = tokens_generated / duration if duration > 0 else 0
                print(f"[Stats] Generated {tokens_generated} tokens in {duration:.2f}s ({speed:.2f} t/s)")
                print(f"[Stats] Total usage: {total_tokens} tokens")
            else:
                 # Fallback if no usage info provided
                 speed = len(full_content) / duration if duration > 0 else 0
                 print(f"[Stats] Generated {len(full_content)} chars in {duration:.2f}s (Approx {speed:.2f} chars/s)")
            
            print(f"[L4 Stream End]\n")
            return full_content

        except Exception as e:
            return f"[Error] Exception during API call: {str(e)}"

File: src/l4_inference/llm_service.ts
import axios, { AxiosResponse } from 'axios';
import { LLMGenerateOptions } from '../types';

/**
 * L4 Inference - LLM Service
 * 大语言模型服务，支持流式输出
 */
export class LLMService {
  private provider: string;
  private apiKey?: string;
  private baseUrl: string;

  constructor(
    provider: string = 'mock',
    apiKey?: string,
    baseUrl: string = 'https://api.siliconflow.cn/v1'
  ) {
    this.provider = provider;
    this.apiKey = apiKey || process.env.SILICONFLOW_API_KEY;
    this.baseUrl = baseUrl;
  }

  /**
   * 生成文本回复
   * @param prompt 用户提示词
   * @param options 生成选项
   * @returns 生成的文本
   */
  async generate(
    prompt: string,
    options: LLMGenerateOptions = {}
  ): Promise<string> {
    const { systemPrompt = '你是一个有用的助手' } = options;

    if (this.provider.startsWith('mock')) {
      return this.generateMock();
    }

    return this.generateApi(prompt, systemPrompt);
  }

  /**
   * Mock 生成器（用于测试）
   */
  private async generateMock(): Promise<string> {
    console.log(`[L4] Generating response using ${this.provider}...`);
    await this.delay(1000);
    return `
(Mock LLM Response from ${this.provider})

**盈亏综述**
账户整体盈利，主要得益于重仓股表现良好。

**归因分析**
- 0700.HK: 腾讯近期发布财报超预期，加上游戏版号发放常态化，带动股价上涨。
- BABA: 阿里拆分消息落地，市场情绪回暖。

**风险提示**
当前持仓科技股占比较高，需注意板块回调风险。

**操作建议**
建议继续持有，可适当关注低估值的红利股以平衡风险。
`;
  }

  /**
   * API 生成器（调用远程大模型）
   */
  private async generateApi(prompt: string, systemPrompt: string): Promise<string> {
    console.log(`[L4] Calling SiliconFlow API for model: ${this.provider}...`);

    if (!this.apiKey) {
      return '[Error] API Key not found. Please set SILICONFLOW_API_KEY environment variable.';
    }

    const url = `${this.baseUrl}/chat/completions`;
    const headers = {
      'Content-Type': 'application/json',
      Authorization: `Bearer ${this.apiKey}`,
    };
    const payload = {
      model: this.provider,
      messages: [
        { role: 'system', content: systemPrompt },
        { role: 'user', content: prompt },
      ],
      enable_thinking: false,
      stream: true,
      stream_options: { include_usage: true },
    };

    try {
      const startTime = Date.now();
      let fullContent = '';
      let usageInfo: { total_tokens?: number; completion_tokens?: number } | null = null;

      console.log(`\n[L4 Stream Start - ${this.provider}]\n`);

      const response: AxiosResponse = await axios.post(url, payload, {
        headers,
        responseType: 'stream',
      });

      return new Promise((resolve, reject) => {
        response.data.on('data', (chunk: Buffer) => {
          const lines = chunk.toString().split('\n');
          for (const line of lines) {
            const trimmedLine = line.trim();
            if (trimmedLine.startsWith('data: ') && trimmedLine !== 'data: [DONE]') {
              const jsonStr = trimmedLine.substring(6);
              try {
                const data = JSON.parse(jsonStr);

                // Handle usage info
                if (data.usage) {
                  usageInfo = data.usage;
                }

                // Handle content delta
                if (data.choices && data.choices[0]) {
                  const delta = data.choices[0].delta;
                  if (delta && delta.content) {
                    process.stdout.write(delta.content);
                    fullContent += delta.content;
                  }
                }
              } catch (e) {
                // Ignore JSON parse errors for incomplete chunks
              }
            }
          }
        });

        response.data.on('end', () => {
          const endTime = Date.now();
          const duration = (endTime - startTime) / 1000;

          console.log('\n');

          // Print Stats
          if (usageInfo) {
            const totalTokens = usageInfo.total_tokens || 0;
            const completionTokens = usageInfo.completion_tokens || 0;
            const tokensGenerated = completionTokens > 0 ? completionTokens : fullContent.length;
            const speed = duration > 0 ? tokensGenerated / duration : 0;

            console.log(`[Stats] Generated ${tokensGenerated} tokens in ${duration.toFixed(2)}s (${speed.toFixed(2)} t/s)`);
            console.log(`[Stats] Total usage: ${totalTokens} tokens`);
          } else {
            const speed = duration > 0 ? fullContent.length / duration : 0;
            console.log(`[Stats] Generated ${fullContent.length} chars in ${duration.toFixed(2)}s (Approx ${speed.toFixed(2)} chars/s)`);
          }

          console.log(`[L4 Stream End]\n`);
          resolve(fullContent);
        });

        response.data.on('error', (error: Error) => {
          reject(new Error(`[Error] Exception during API call: ${error.message}`));
        });
      });
    } catch (error) {
      if (axios.isAxiosError(error)) {
        return `[Error] API request failed: ${error.response?.status} - ${error.response?.data}`;
      }
      return `[Error] Exception during API call: ${(error as Error).message}`;
    }
  }

  /**
   * 延迟辅助函数
   */
  private delay(ms: number): Promise<void> {
    return new Promise((resolve) => setTimeout(resolve, ms));
  }
}
